lib2/provider/awssamlprovider.go

```diff
@@ -0,0 +1,450 @@
+// manages auth sessions for Okta applications
+package provider
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/segmentio/aws-okta/internal/sessioncache"
+	"github.com/segmentio/aws-okta/lib"
+
+	"github.com/99designs/keyring"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	aws_session "github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+	log "github.com/sirupsen/logrus"
+	// use xerrors until 1.13 is stable/oldest supported version
+	"golang.org/x/xerrors"
+)
+
+const (
+	MaxSessionDuration    = time.Hour * 24 * 90
+	MinSessionDuration    = time.Minute * 15
+	MinAssumeRoleDuration = time.Minute * 15
+	MaxAssumeRoleDuration = time.Hour * 12
+
+	DefaultSessionDuration    = time.Hour * 4
+	DefaultAssumeRoleDuration = time.Minute * 15
+)
+
+type SessionCacheInterface interface {
+	Get(sessioncache.Key) (*sessioncache.Session, error)
+	Put(sessioncache.Key, *sessioncache.Session) error
+}
+
+type OktaClient interface {
+	AuthenticateUser() error
+	GetSessionToken() string
+	SaveSessionCookie() error
+	Request(method string, path string, queryParams url.Values, data []byte, format string, followRedirects bool) (*http.Response, error)
+	GetURL(string) (*url.URL, error)
+}
+
+type AwsSamlProvider struct {
+	credentials.Expiry
+	AwsSamlProviderOptions
+	oktaClient             OktaClient
+	keyring                keyring.Keyring
+	profileARN             string
+	oktaAwsSAMLUrl         string
+	oktaAccountName        string
+	awsRegion              string
+	profile                string
+	Expires                time.Time
+	sessions               SessionCacheInterface
+	defaultRoleSessionName string
+}
+
+type AwsSamlProviderOptions struct {
+	SessionDuration    time.Duration
+	AssumeRoleDuration time.Duration
+	ExpiryWindow       time.Duration
+	Profiles           lib.Profiles
```

Comment:
`Profiles` needs more structure and methods. Looking up things in a map is super lame

Reply:
I this needs changes, could we tackle this as part of another issue?

Reply:
Well it depends I suppose. After this refactor, are we expecting this to be `v1` of the API or still `v0`? You're going to be the consumer, so if we stick with `v0` and the API changes in a minor release, would that be painful for you?

Reply:
I think we're good to go ahead with a `v1` release.